Cmake error on MPI

Hi, I get the below error when I run the cmake . command from within the git hub cloned version of espresso++ on ubuntu 18.04. Log error file is also reported afterwards.
There's something wrong with my MPI settings I believe, but I have no idea and struggling track back in cmake what's wrong. All mpi fortran libs and compilers should be there.
Please help! Thank you very much!

mariano@mariano-HP-Pavilion-15-Notebook-PC:~/Programs/espressopp$ cmake .
-- Could NOT find MPI_CXX (missing: MPI_CXX_WORKS) 
CMake Error at /home/mariano/cmake-3.17.0-Linux-x86_64/share/cmake-3.17/Modules/FindPackageHandleStandardArgs.cmake:164 (message):
  Could NOT find MPI (missing: MPI_CXX_FOUND) (found version "3.1")

      Reason given by package: MPI component 'Fortran' was requested, but language Fortran is not enabled.  

Call Stack (most recent call first):
  /home/mariano/cmake-3.17.0-Linux-x86_64/share/cmake-3.17/Modules/FindPackageHandleStandardArgs.cmake:445 (_FPHSA_FAILURE_MESSAGE)
  /home/mariano/cmake-3.17.0-Linux-x86_64/share/cmake-3.17/Modules/FindMPI.cmake:1717 (find_package_handle_standard_args)
  CMakeLists.txt:70 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/mariano/Programs/espressopp/CMakeFiles/CMakeError.log".


--------------------------------------------------------------------------------
**Error log file** 

The MPI test test_mpi for CXX in mode normal failed to compile with the following output:
Change Dir: /home/mariano/Programs/espressopp/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_14f96/fast && /usr/bin/make -f CMakeFiles/cmTC_14f96.dir/build.make CMakeFiles/cmTC_14f96.dir/build
make[1]: Entering directory '/home/mariano/Programs/espressopp/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_14f96.dir/test_mpi.cpp.o
/usr/bin/c++   -isystem /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -isystem /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/opal/mca/event/libevent2022/libevent -isystem /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/opal/mca/event/libevent2022/libevent/include -isystem /usr/lib/x86_64-linux-gnu/openmpi/include -isystem /usr/include/mpich  -pthread -std=c++11 -o CMakeFiles/cmTC_14f96.dir/test_mpi.cpp.o -c /home/mariano/Programs/espressopp/CMakeFiles/FindMPI/test_mpi.cpp
Linking CXX executable cmTC_14f96
/home/mariano/cmake-3.17.0-Linux-x86_64/bin/cmake -E cmake_link_script CMakeFiles/cmTC_14f96.dir/link.txt --verbose=1
/usr/bin/c++       -Wl,-Bsymbolic-functions -Wl,-z,relro CMakeFiles/cmTC_14f96.dir/test_mpi.cpp.o  -o cmTC_14f96  /usr/lib/x86_64-linux-gnu/libmpichcxx.so /usr/lib/x86_64-linux-gnu/libmpich.so 
CMakeFiles/cmTC_14f96.dir/test_mpi.cpp.o: In function `MPI::Intracomm::Intracomm()':
test_mpi.cpp:(.text._ZN3MPI9IntracommC2Ev[_ZN3MPI9IntracommC5Ev]+0x14): undefined reference to `MPI::Comm::Comm()'
CMakeFiles/cmTC_14f96.dir/test_mpi.cpp.o: In function `MPI::Intracomm::Intracomm(ompi_communicator_t*)':
test_mpi.cpp:(.text._ZN3MPI9IntracommC2EP19ompi_communicator_t[_ZN3MPI9IntracommC5EP19ompi_communicator_t]+0x28): undefined reference to `MPI::Comm::Comm()'
test_mpi.cpp:(.text._ZN3MPI9IntracommC2EP19ompi_communicator_t[_ZN3MPI9IntracommC5EP19ompi_communicator_t]+0x4d): undefined reference to `ompi_mpi_comm_null'
test_mpi.cpp:(.text._ZN3MPI9IntracommC2EP19ompi_communicator_t[_ZN3MPI9IntracommC5EP19ompi_communicator_t]+0x88): undefined reference to `ompi_mpi_comm_null'
CMakeFiles/cmTC_14f96.dir/test_mpi.cpp.o: In function `MPI::Cartcomm::Cartcomm(ompi_communicator_t* const&)':
test_mpi.cpp:(.text._ZN3MPI8CartcommC2ERKP19ompi_communicator_t[_ZN3MPI8CartcommC5ERKP19ompi_communicator_t]+0x54): undefined reference to `ompi_mpi_comm_null'
test_mpi.cpp:(.text._ZN3MPI8CartcommC2ERKP19ompi_communicator_t[_ZN3MPI8CartcommC5ERKP19ompi_communicator_t]+0xa3): undefined reference to `ompi_mpi_comm_null'
CMakeFiles/cmTC_14f96.dir/test_mpi.cpp.o: In function `MPI::Graphcomm::Graphcomm(ompi_communicator_t* const&)':
test_mpi.cpp:(.text._ZN3MPI9GraphcommC2ERKP19ompi_communicator_t[_ZN3MPI9GraphcommC5ERKP19ompi_communicator_t]+0x54): undefined reference to `ompi_mpi_comm_null'
CMakeFiles/cmTC_14f96.dir/test_mpi.cpp.o:test_mpi.cpp:(.text._ZN3MPI9GraphcommC2ERKP19ompi_communicator_t[_ZN3MPI9GraphcommC5ERKP19ompi_communicator_t]+0xa3): more undefined references to `ompi_mpi_comm_null' follow
CMakeFiles/cmTC_14f96.dir/test_mpi.cpp.o: In function `MPI::Op::Init(void (*)(void const*, void*, int, MPI::Datatype const&), bool)':
test_mpi.cpp:(.text._ZN3MPI2Op4InitEPFvPKvPviRKNS_8DatatypeEEb[_ZN3MPI2Op4InitEPFvPKvPviRKNS_8DatatypeEEb]+0x26): undefined reference to `ompi_mpi_cxx_op_intercept'
test_mpi.cpp:(.text._ZN3MPI2Op4InitEPFvPKvPviRKNS_8DatatypeEEb[_ZN3MPI2Op4InitEPFvPKvPviRKNS_8DatatypeEEb]+0x45): undefined reference to `ompi_op_set_cxx_callback'
collect2: error: ld returned 1 exit status
CMakeFiles/cmTC_14f96.dir/build.make:105: recipe for target 'cmTC_14f96' failed
make[1]: *** [cmTC_14f96] Error 1
make[1]: Leaving directory '/home/mariano/Programs/espressopp/CMakeFiles/CMakeTmp'
Makefile:138: recipe for target 'cmTC_14f96/fast' failed
make: *** [cmTC_14f96/fast] Error 2

Hi,

so I tried to play adding/removing openmpi and mpich and the error persist in any case.
Then in tried to edit the CmakeList file adding fortran as language, and cmake went through (!) but I get another error:
-- Could NOT find MPI_CXX (missing: MPI_CXX_WORKS) 
CMake Error at /home/mariano/cmake-3.17.0-Linux-x86_64/share/cmake-3.17/Modules/FindPackageHandleStandardArgs.cmake:164 (message):
  Could NOT find MPI (missing: MPI_CXX_FOUND) (found version "3.1")
Call Stack (most recent call first):
  /home/mariano/cmake-3.17.0-Linux-x86_64/share/cmake-3.17/Modules/FindPackageHandleStandardArgs.cmake:445 (_FPHSA_FAILURE_MESSAGE)
  /home/mariano/cmake-3.17.0-Linux-x86_64/share/cmake-3.17/Modules/FindMPI.cmake:1717 (find_package_handle_standard_args)
  CMakeLists.txt:71 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/mariano/Programs/espressopp/CMakeFiles/CMakeOutput.log".
See also "/home/mariano/Programs/espressopp/CMakeFiles/CMakeError.log".

so now it cannot find MPI_CXX or MPI at all...

Then, I tried to use spack to install it, and it went smooth till installing boost and failed immediately afterwards. I have created an issue on spack on this issue also.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,12 +1,9 @@
 cmake_minimum_required(VERSION 3.10)
 
-project(espressopp)
+project(espressopp LANGUAGES CXX)
 
 # Cmake modules/macros are in a subdirectory to keep this file cleaner
 set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
-
-enable_language(C)
-enable_language(CXX)
 
 enable_testing()
 
@@ -67,8 +64,7 @@ endif(VAMPIRTRACE)
 #Process MPI settings
 ########################################################################
 
-find_package(MPI REQUIRED)
-include_directories(${MPI_INCLUDE_PATH})
+find_package(MPI REQUIRED COMPONENTS CXX)
 
 ########################################################################
 #Process FFTW3 settings

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -12,7 +12,6 @@ enable_testing()
 # these need ot be done early (before further tests).
 #####################################################################
 
-include(CheckCCompilerFlag)
 include(CheckCXXCompilerFlag)
 
 # Check for C++11 standard
@@ -135,9 +134,6 @@ mark_as_advanced(PYTHON_INCLUDE_PATH PYTHON_LIBRARIES)
 ########################################################################
 # Basic system tests (standard libraries, headers, functions, types)   #
 ########################################################################
-include(CheckIncludeFiles)
-check_include_files("sys/time.h;sys/resource.h;unistd.h" CMAKE_HEADERS)
-
 list(APPEND EXTRA_LIBRARIES ${CMAKE_DL_LIBS})
 
 # this conflicts with mpi4py-1.3 (is used to check for MPI version there)

--- a/contrib/boost/CMakeLists.txt
+++ b/contrib/boost/CMakeLists.txt
@@ -44,7 +44,7 @@ endif()
 add_library(espressopp_boost STATIC ${BOOST_SOURCES})
 # Shared libboost library for unit tests.
 add_library(espressopp_boost_test SHARED ${BOOST_SOURCES_TEST})
-target_link_libraries(espressopp_boost ${PYTHON_LIBRARIES} ${VAMPIRTRACE_LIBRARIES} ${MPI_LIBRARIES} ${RT_LIBRARIES})
+target_link_libraries(espressopp_boost ${PYTHON_LIBRARIES} ${VAMPIRTRACE_LIBRARIES} MPI::MPI_CXX ${RT_LIBRARIES})
 if( CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64" )
   set_target_properties(espressopp_boost PROPERTIES COMPILE_FLAGS "-fPIC")
   set_target_properties(espressopp_boost_test PROPERTIES COMPILE_FLAGS "-fPIC")

--- a/contrib/boost/CMakeLists.txt
+++ b/contrib/boost/CMakeLists.txt
@@ -45,6 +45,7 @@ add_library(espressopp_boost STATIC ${BOOST_SOURCES})
 # Shared libboost library for unit tests.
 add_library(espressopp_boost_test SHARED ${BOOST_SOURCES_TEST})
 target_link_libraries(espressopp_boost ${PYTHON_LIBRARIES} ${VAMPIRTRACE_LIBRARIES} MPI::MPI_CXX ${RT_LIBRARIES})
+target_link_libraries(espressopp_boost_test MPI::MPI_CXX)
 if( CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64" )
   set_target_properties(espressopp_boost PROPERTIES COMPILE_FLAGS "-fPIC")
   set_target_properties(espressopp_boost_test PROPERTIES COMPILE_FLAGS "-fPIC")

--- a/contrib/mpi4py/CMakeLists.txt
+++ b/contrib/mpi4py/CMakeLists.txt
@@ -1,3 +1,5 @@
+enable_language(C)
+
 set(MPI4PY_DIR mpi4py-2.0.0) 
 set(MPI4PY_SOURCES ${MPI4PY_DIR}/src/MPI.c)
 set(DL_SOURCES ${MPI4PY_DIR}/src/dynload.c)
@@ -5,11 +7,11 @@ set(DL_SOURCES ${MPI4PY_DIR}/src/dynload.c)
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/${MPI4PY_DIR}/src)
 
 add_library(MPI ${MPI4PY_SOURCES})
-target_link_libraries(MPI ${PYTHON_LIBRARIES} ${VAMPIRTRACE_LIBRARIES} ${MPI_LIBRARIES})
+target_link_libraries(MPI ${PYTHON_LIBRARIES} ${VAMPIRTRACE_LIBRARIES} MPI::MPI_CXX)
 set_target_properties(MPI PROPERTIES PREFIX "" SUFFIX ".so")
 
 add_library(dl ${DL_SOURCES})
-target_link_libraries(dl ${PYTHON_LIBRARIES} ${VAMPIRTRACE_LIBRARIES} ${MPI_LIBRARIES})
+target_link_libraries(dl ${PYTHON_LIBRARIES} ${VAMPIRTRACE_LIBRARIES} MPI::MPI_CXX)
 set_target_properties(dl PROPERTIES PREFIX "" SUFFIX ".so")
 
 install(TARGETS MPI LIBRARY DESTINATION ${PYTHON_INSTDIR}/mpi4py ARCHIVE DESTINATION ${PYTHON_INSTDIR}/mpi4py)

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -48,7 +48,7 @@ endif (WITH_RC_FILES)
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/include)
 
 add_library(_espressopp ${ESPRESSO_SOURCES})
-target_link_libraries(_espressopp ${Boost_LIBRARIES} ${PYTHON_LIBRARIES} ${MPI_LIBRARIES} ${FFTW3_LIBRARIES} ${VAMPIRTRACE_LIBRARIES})
+target_link_libraries(_espressopp ${Boost_LIBRARIES} ${PYTHON_LIBRARIES} MPI::MPI_CXX ${FFTW3_LIBRARIES} ${VAMPIRTRACE_LIBRARIES})
 if(WITH_XTC)
   target_link_libraries(_espressopp ${GROMACS_LIBRARIES})
 endif()

--- a/testsuite/CMakeLists.txt
+++ b/testsuite/CMakeLists.txt
@@ -33,7 +33,7 @@ foreach(DIR ${ALL_DIRS})
                 endif()
                 add_test(${TEST_NAME} ${CMAKE_CURRENT_BINARY_DIR}/${TEST_NAME})
                 set_tests_properties(${TEST_NAME} PROPERTIES ENVIRONMENT "${TEST_ENV}")
-                target_link_libraries(${TEST_NAME} ${PYTHON_LIBRARIES} ${MPI_LIBRARIES} ${FFTW3_LIBRARIES} _espressopp ${Boost_LIBRARIES})
+                target_link_libraries(${TEST_NAME} ${PYTHON_LIBRARIES} MPI::MPI_CXX ${FFTW3_LIBRARIES} _espressopp ${Boost_LIBRARIES})
             endforeach()
 
             file(GLOB TESTS_PY "${DIR}/test*.py")

